### PR TITLE
Products by Attribute: Update block to use API, shortcode rendering

### LINKED ISSUE #361: Display "Products by Attribute" on front end of site
A sub-issue for #243 – We need a way to render the Products by Attribute block on the front end. We might have a combination of multiple attribute terms across attributes.  For example, we might have attributes Color and Material, and we need to be able to show combinations like: "red and cotton", "blue or linen", "red or blue", etc.

Currently we render all blocks using [the `[products ]` shortcode,](https://docs.woocommerce.com/document/woocommerce-shortcodes/#content-product-attributes) but we can only pass one `attribute` slug. We could update the shortcode.

Or perhaps this is [the first dynamic block,](https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/block-tutorial/creating-dynamic-blocks/) where we use a `render_callback` to display the products? We could query the DB for a set of product IDs, then render the shortcode with those IDs (to keep the themeable output the same).

Or some other idea?

### PR DESCRIPTION
**Pull request by ryelle**
Fixes #361 – Uses the new design to load attribute types, then once selected it will load the terms in that attribute for selection. Once selected, the block uses the updated API to pull just the products matching those attribute terms, and on save it converts that to the shortcode.

#### Accessibility

Functionally this works, but I want to tweak the screen reader experience some more.

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

Note: I know the design isn't quite there, I'll put in a follow-up PR to fix the CSS & screen reader issues. This is just to make sure the functionality works as expected. I'm not tagging this as design review, but I would appreciate a "flow review" from @jameskoster 🙂 

screen shot 2019-02-06 at 6 34 34 pm

### How to test the changes in this Pull Request:

1. Add a Products by Attribute block
2. Select different sets of attributes
3. Expect: The products that display should match the selected attributes

### CODE REVIEW

File: assets/js/components/product-attribute-control/index.js [hunk 1]
```diff
@@ -22,38 +22,84 @@ class ProductAttributeControl extends Component {
...
+			} ),
+		} )
+			.then( ( terms ) => {
+				terms = terms.map( ( term ) => ( { ...term, parent: attribute } ) );
+				this.setState( ( { list } ) => ( {
+					list: uniqBy( [ ...list, ...terms ], 'id' ),
+					termsLoading: false,
+				} ) );
+			} )
+			.catch( () => {
```
Comment (joshuatf) [inline assets/js/components/product-attribute-control/index.js]: I see this is the same in other areas as well, but curious if we ever want to do anything should we catch an error.  May be a task for another PR though.

File: assets/js/components/product-attribute-control/index.js [hunk 2]
```diff
@@ -22,38 +22,84 @@ class ProductAttributeControl extends Component {
...
+		}
+
+		apiFetch( {
+			path: addQueryArgs( `/wc-pb/v3/products/attributes/${ attribute }/terms`, {
+				per_page: -1,
+			} ),
+		} )
+			.then( ( terms ) => {
+				terms = terms.map( ( term ) => ( { ...term, parent: attribute } ) );
+				this.setState( ( { list } ) => ( {
```
Comment (joshuatf) [inline assets/js/components/product-attribute-control/index.js]: This is working well as is, but I was also thinking that storing all attribute terms in an object might remove the delay when toggling back and forth between attributes instead of refetching on change.

```js
attributes: {
  color: [ 'red', 'blue' ],
  size: [ 'medium', 'large' ],
};
```
What do you think?

**Review by joshuatf — APPROVED** [review]
This is testing well and looks good from a dev perspective.  Left a comment about possibly storing attributes in state, but pending design review, things look good to merge.

Only other thought I had is if we need to make any changes to attributes without any terms (disabling them to avoid confusion or possibly listing a '0' count next to the attribute name). /cc @jameskoster 

screen shot 2019-02-07 at 4 33 00 pm

**Review by jameskoster — COMMENTED** [review]
Looking good! I tweaked the CSS a little. 

Only thing that feels a little off is the pause when loading the terms. Would it be possible to have a loading state for that, similar to the concept [here](https://www.figma.com/proto/NzFuASLQS8wqTXSRNCj1A198/Featured-Product?node-id=1%3A2&viewport=501%2C598%2C0.301967&scaling=min-zoom)?

File: assets/js/components/product-attribute-control/index.js (hunk 1 above)
Comment (ryelle) [reply assets/js/components/product-attribute-control/index.js]: Yeah, we've got a catch-all issue to look at this pattern: https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/380

File: assets/js/components/product-attribute-control/index.js (hunk 2 above)
Comment (ryelle) [reply assets/js/components/product-attribute-control/index.js]: That's a good idea, but I think I'll handle that in a follow up PR.

**Comment by ryelle** [comment]
> Only other thought I had is if we need to make any changes to attributes without any terms (disabling them to avoid confusion or possibly listing a '0' count next to the attribute name).

As I mentioned, this isn't following the design yet– each item will have a "X items" count next to it, but thanks for reminding me that an attribute could have 0 items.

> Would it be possible to have a loading state for that, similar to the concept here

Yep – just wanted to get the functionality done first.

**Comment by jameskoster** [comment]
On second thoughts, we should probably use the core "loading" pattern, assuming there's a component for this. https://cloudup.com/cohBzS65Awf

**Comment by ryelle** [comment]
@jameskoster it's not clear from your screencast - do you mean the spinner, like appears when you toggle into the edit mode on the Products by Category or Featured Product views? Or something else?

**Comment by ryelle** [comment]
@jameskoster Did you mean for the style updates to affect all of these SearchListControls? Products by Category looks like this now:

screen shot 2019-02-07 at 11 04 57 am

**Comment by jameskoster** [comment]
> @jameskoster it's not clear from your screencast - do you mean the spinner, like appears when you toggle into the edit mode on the Products by Category or Featured Product views? Or something else?

Yup, the spinner.

> @jameskoster Did you mean for the style updates to affect all of these SearchListControls? Products by Category looks like this now:

Yeah I think it makes sense to group parents/children together.
